sdk/tools/xml2sdb/xml2sdb.cpp

```diff
@@ -700,65 +740,70 @@ bool Database::toSdb(LPCWSTR path)
     EndWriteListTag(pdb, tidLibrary);
     if (!WriteGeneric(pdb, Layers, *this))
         return false;
-    if (!WriteGeneric(pdb, Exes, *this))
+    if (!WriteGeneric(pdb, Exes, *this, TAG_EXE))
+        return false;
+
+    if (!WriteGeneric(pdb, KDrivers, *this, TAG_KDRIVER))
+        return false;
+    if (!WriteGeneric(pdb, KDevices, *this, TAG_KDEVICE))
         return false;
+    if (!WriteGeneric(pdb, KShims, *this))
+        return false;
+
     EndWriteListTag(pdb, tidDatabase);
 
     SdbCloseDatabaseWrite(pdb);
     return true;
 }
 
-static void InsertTagid(const sdbstring& name, TAGID tagid, std::map<sdbstring, TAGID>& lookup, const char* type)
+bool iequals(const std::string& a, const std::string& b)
 {
-    sdbstring nameLower = name;
-    std::transform(nameLower.begin(), nameLower.end(), nameLower.begin(), ::tolower);
-    if (lookup.find(nameLower) != lookup.end())
-    {
-        std::string nameA(name.begin(), name.end());
-        SHIM_WARN("%s '%s' redefined\n", type, nameA.c_str());
-        return;
-    }
-    lookup[nameLower] = tagid;
-}
-
-static TAGID FindTagid(const sdbstring& name, const std::map<sdbstring, TAGID>& lookup)
-{
-    sdbstring nameLower = name;
-    std::transform(nameLower.begin(), nameLower.end(), nameLower.begin(), ::tolower);
-    std::map<sdbstring, TAGID>::const_iterator it = lookup.find(nameLower);
-    if (it == lookup.end())
-        return 0;
-    return it->second;
-}
-
-void Database::InsertShimTagid(const sdbstring& name, TAGID tagid)
-{
-    InsertTagid(name, tagid, KnownShims, "Shim");
-}
-
-TAGID Database::FindShimTagid(const sdbstring& name)
-{
-    return FindTagid(name, KnownShims);
-}
-
-void Database::InsertPatchTagid(const sdbstring& name, TAGID tagid)
-{
-    InsertTagid(name, tagid, KnownPatches, "Patch");
+    unsigned int sz = a.size();
+    if (b.size() != sz)
+        return false;
+    for (unsigned int i = 0; i < sz; ++i)
+        if (tolower(a[i]) != tolower(b[i]))
+            return false;
+    return true;
 }
 
-TAGID Database::FindPatchTagid(const sdbstring& name)
+TAGID Database::FindShimTagid(const std::string& name)
 {
-    return FindTagid(name, KnownPatches);
+    for (const auto& it : Library.Shims)
+    {
+        if (iequals(name, it.Name))
+        {
+            return it.Tagid;
+        }
+    }
+    assert(false);
```

Comment:
These are "it really should be impossible to get here" asserts, not "invalid user input" asserts, right?

Reply:
this one was mainly used while writing this, but it could mean invalid user input now.

Reply:
I'll add some nice error messages in this tool on the next PR